### PR TITLE
Add support for configuring gRPC max-connection-age for the kiam server

### PR DESCRIPTION
**Pull request by stefansedich**
This PR adds support for configuring the `max-connection` related options for the kiam server.

Closes https://github.com/uswitch/kiam/issues/433

### CODE REVIEW

**Comment by stefansedich** [comment]
@pingles curios on your thoughts here, I added support for all of the `max-connection` options, the one thing I could not see how to do cleaner was to default to infinity for the values.

**Comment by pingles** [comment]
Thanks. Not sure how to better handle defaults, although perhaps the infinity defaults would be better to be limited anyway?

The failure mode occurs when a server fails to resolve addresses (or server addresses are removed) but the load-balance won't re-fill the pool unless a connection change occurs. So the gRPC team's advice was to set those settings to ensure it triggers a reconnection.

I wanted to get some tests to try and prove the failing behaviour today, and managed to get something rolling with https://github.com/uswitch/kiam/pull/441.

**Comment by stefansedich** [comment]
> Thanks. Not sure how to better handle defaults, although perhaps the infinity defaults would be better to be limited anyway?
> 
> The failure mode occurs when a server fails to resolve addresses (or server addresses are removed) but the load-balance won't re-fill the pool unless a connection change occurs. So the gRPC team's advice was to set those settings to ensure it triggers a reconnection.
> 
> I wanted to get some tests to try and prove the failing behaviour today, and managed to get something rolling with #441.

I was not sure of a good default here, but you are right we can do something more reasonable for "infinity" I am sure 1 month is even enough here? or we could pick a reasonable default that we know works whether that be 1 minute or something else we decide on.

**Comment by pingles** [comment]
I would've thought something on the order of 15 minutes is a good starting point. https://github.com/pomerium/pomerium set it to 5 minutes.

**Comment by stefansedich** [comment]
> I would've thought something on the order of 15 minutes is a good starting point. https://github.com/pomerium/pomerium set it to 5 minutes.

Done!

File: cmd/kiam/server.go [hunk 1]
```diff
@@ -59,6 +59,11 @@ func (o *serverOptions) bind(parser parser) {
 	parser.Flag("session-refresh", "How soon STS Tokens should be refreshed before their expiration.").Default("5m").DurationVar(&o.SessionRefresh)
 	parser.Flag("assume-role-arn", "IAM Role to assume before processing requests").Default("").StringVar(&o.AssumeRoleArn)
 	parser.Flag("region", "AWS Region to use for regional STS calls (e.g. us-west-2). Defaults to the global endpoint.").Default("").StringVar(&o.Region)
+	parser.Flag("grpc-keepalive-time-duration", "gRPC keepalive time").Default("10s").DurationVar(&o.KeepaliveParams.Time)
```
Comment (stefansedich) [inline cmd/kiam/server.go:62]: I took the defaults for the time/timeout from the existing agent options.

File: cmd/kiam/server.go [hunk 2]
```diff
@@ -59,6 +59,11 @@ func (o *serverOptions) bind(parser parser) {
...
+	parser.Flag("grpc-keepalive-timeout-duration", "gRPC keepalive timeout").Default("2s").DurationVar(&o.KeepaliveParams.Timeout)
+	parser.Flag("grpc-max-connection-idle-duration", "gRPC max connection idle").Default("15m").DurationVar(&o.KeepaliveParams.MaxConnectionIdle)
```
Comment (stefansedich) [inline cmd/kiam/server.go:64]: Defaulted max-connection options to 15m as discussed.

**Comment by pingles** [comment]
Thanks @stefansedich. I think some of those defaults may need to increase but let me pull the code and check.

**Comment by pingles** [comment]
Actually, looks ok to me. I think it'd be good to get some real-world usage and see how it behaves but seems sensible to me to start with these.

Thanks again.

**Review by pingles — APPROVED** [review]

**Pull request merged**